File: frontend/lib/schema.ts
/**
 * TypeScript-Typen und Helfer für das schema-getriebene Fragebogen-Format
 * (Format-Version 2). Spiegel von backend/questionnaires/schema.py.
 */

export type QuestionType =
  | "yes_no"
  | "choice"
  | "multi_choice"
  | "text"
  | "textarea"
  | "ess_matrix"
  | "consent";

export interface QuestionOption {
  value: string;
  label: string;
}

export interface Followup {
  id: string;
  type: "text" | "textarea";
  label: string;
  /** Antwortwert der Elternfrage, bei dem das Detailfeld erscheint (Default "yes") */
  when?: string;
  required?: boolean;
}

export type ShowIf =
  | { id: string; in: unknown[]; not_in?: never }
  | { id: string; not_in: unknown[]; in?: never };

export interface EssItem {
  id: string;
  label: string;
}

export interface Question {
  id: string;
  type: QuestionType;
  label: string;
  hint?: string;
  required?: boolean;
  error?: string;
  options?: QuestionOption[];
  followup?: Followup;
  show_if?: ShowIf;
  /** Nur ess_matrix: die 8 ESS-Zeilen (ess_1..ess_8) */
  items?: EssItem[];
}

export interface Section {
  id: string;
  title: string;
  subtitle?: string;
  questions: Question[];
  pdf_note?: string;
}

export interface Schema {
  version: number;
  title?: string;
  basis?: string;
  sections: Section[];
}

export type Answers = Record<string, unknown>;

/** True, wenn das Schema das strukturierte v2-Format hat (sections als Objekte). */
export function isV2Schema(schema: unknown): schema is Schema {
  if (!schema || typeof schema !== "object") return false;
  const sections = (schema as { sections?: unknown }).sections;
  return (
    Array.isArray(sections) &&
    sections.length > 0 &&
    typeof sections[0] === "object" &&
    sections[0] !== null
  );
}

/**
 * Wertet show_if gegen die gegebenen Antworten aus
 * (exakt wie backend/questionnaires/schema.py::is_visible).
 */
export function isVisible(question: Question, answers: Answers): boolean {
  const cond = question.show_if;
  if (!cond) return true;
  const value = answers[cond.id];
  if ("in" in cond && cond.in) {
    return cond.in.includes(value as never);
  }
  if ("not_in" in cond && cond.not_in) {
    // Nur sichtbar, wenn eine Antwort vorhanden ist UND sie nicht in der Liste steht
    return value !== undefined && value !== null && !cond.not_in.includes(value as never);
  }
  return true;
}

/** Alle aktuell sichtbaren Fragen einer Sektion. */
export function visibleQuestions(section: Section, answers: Answers): Question[] {
  return section.questions.filter((q) => isVisible(q, answers));
}

/**
 * Antwortwert → Anzeigetext (für Print/PDF), analog
 * backend/questionnaires/schema.py::answer_display.
 */
export function answerDisplay(question: Question, value: unknown): string {
  if (value === null || value === undefined || value === "") return "—";
  const qtype = question.type;
  if (qtype === "yes_no") {
    return value === "yes" ? "Ja" : value === "no" ? "Nein" : "—";
  }
  if (qtype === "consent") {
    return value === true ? "Ja" : "Nein";
  }
  if (qtype === "choice") {
    const opt = (question.options ?? []).find((o) => o.value === value);
    return opt ? opt.label : String(value);
  }
  if (qtype === "multi_choice" && Array.isArray(value)) {
    const byValue = new Map((question.options ?? []).map((o) => [o.value, o.label]));
    const labels = value.map((v) => byValue.get(v as string) ?? String(v));
    return labels.length > 0 ? labels.join(", ") : "—";
  }
  return String(value);
}

File: backend/questionnaires/schema.py
"""
Schema-getriebener Fragebogen: Hilfsfunktionen für Validierung und Rendering.

Das Schema liegt als schema_json am QuestionnaireTemplate (Format-Version 2):

{
  "version": 2,
  "title": "...",
  "basis": "...",
  "sections": [
    {
      "id": "fahrprofil", "title": "...", "subtitle": "...",
      "questions": [
        {"id": "...", "type": "yes_no|choice|multi_choice|text|textarea|ess_matrix|consent",
         "label": "...", "hint": "...", "required": true,
         "options": [{"value": "...", "label": "..."}],       # choice/multi_choice
         "followup": {"id": "...", "type": "text|textarea", "label": "...",
                       "when": "yes", "required": false},       # optionales Detailfeld
         "show_if": {"id": "andere_frage", "in": [...]} |
                    {"id": "andere_frage", "not_in": [...]}     # bedingte Sichtbarkeit
        }
      ]
    }
  ]
}

Antwortformat (answers_json): {frage_id: wert}. yes_no → "yes"/"no",
choice → Options-value, multi_choice → Liste von values, consent → true,
ess_matrix → ess_1..ess_8 als int 0-3 (direkt auf oberster Ebene).
"""

ESS_KEYS = [f"ess_{i}" for i in range(1, 9)]

MAX_TEXT_LENGTH = 2000


def is_v2_schema(schema):
    """True, wenn schema_json das strukturierte Format (sections als Objekte) hat."""
    sections = (schema or {}).get("sections")
    return bool(sections) and isinstance(sections[0], dict)


def iter_questions(schema):
    """Alle (section, question)-Paare in Dokumentreihenfolge."""
    for section in schema.get("sections", []):
        for question in section.get("questions", []):
            yield section, question


def is_visible(question, answers):
    """Wertet show_if gegen die gegebenen Antworten aus."""
    cond = question.get("show_if")
    if not cond:
        return True
    value = answers.get(cond.get("id"))
    if "in" in cond:
        return value in cond["in"]
    if "not_in" in cond:
        return value is not None and value not in cond["not_in"]
    return True


def _clean_text(value):
    return str(value).strip()[:MAX_TEXT_LENGTH]


def _option_values(question):
    return [o.get("value") for o in question.get("options", [])]


def validate_answers(schema, data):
    """
    Validiert eingereichte Antworten gegen das Schema.

    Returns (cleaned, errors):
      cleaned – Whitelist der bekannten Antworten (inkl. ess_total/ess_band),
      errors  – {frage_id: meldung}; leer bei Erfolg.
    """
    cleaned = {}
    errors = {}

    def require(qid, label):
        errors[qid] = f"„{label}“ ist erforderlich."

    def handle_followup(question):
        follow = question.get("followup")
        if not follow:
            return
        when = follow.get("when", "yes")
        if data.get(question["id"]) != when:
            return
        raw = data.get(follow["id"])
        if raw is None or _clean_text(raw) == "":
            if follow.get("required"):
                require(follow["id"], follow.get("label", follow["id"]))
            return
        cleaned[follow["id"]] = _clean_text(raw)

    for _section, q in iter_questions(schema):
        qid = q["id"]
        qtype = q.get("type")
        label = q.get("label", qid)
        required = bool(q.get("required"))

        if not is_visible(q, data):
            continue

        if qtype == "ess_matrix":
            for key in ESS_KEYS:
                raw = data.get(key)
                try:
                    value = int(raw)
                except (TypeError, ValueError):
                    errors[key] = "Bitte einen Wert von 0 bis 3 auswählen."
                    continue
                if not 0 <= value <= 3:
                    errors[key] = "Bitte einen Wert von 0 bis 3 auswählen."
                    continue
                cleaned[key] = value
            continue

        raw = data.get(qid)

        if qtype == "yes_no":
            if raw in ("yes", "no"):
                cleaned[qid] = raw
                handle_followup(q)
            elif required:
                require(qid, label)

        elif qtype == "choice":
            if raw in _option_values(q):
                cleaned[qid] = raw
                handle_followup(q)
            elif required:
                require(qid, label)

        elif qtype == "multi_choice":
            allowed = set(_option_values(q))
            if isinstance(raw, list):
                values = [v for v in raw if v in allowed]
                if values:
                    cleaned[qid] = values
                elif required:
                    require(qid, label)
            elif required:
                require(qid, label)

        elif qtype in ("text", "textarea"):
            if raw is not None and _clean_text(raw) != "":
                cleaned[qid] = _clean_text(raw)
            elif required:
                require(qid, label)

        elif qtype == "consent":
            if raw is True:
                cleaned[qid] = True
            else:
                errors[qid] = q.get(
                    "error", "Bitte bestätigen Sie diese Erklärung."
                )

    if not errors and all(k in cleaned for k in ESS_KEYS):
        total = sum(cleaned[k] for k in ESS_KEYS)
        cleaned["ess_total"] = total
        cleaned["ess_band"] = ess_band(total)

    return cleaned, errors


def ess_band(total):
    if total >= 16:
        return "ausgeprägt"
    if total >= 10:
        return "erhöht"
    return "normal"


def answer_display(question, value):
    """Antwortwert → Anzeigetext (für PDF/Print)."""
    if value is None or value == "":
        return "—"
    qtype = question.get("type")
    if qtype == "yes_no":
        return {"yes": "Ja", "no": "Nein"}.get(value, "—")
    if qtype == "consent":
        return "Ja" if value is True else "Nein"
    if qtype in ("choice",):
        for opt in question.get("options", []):
            if opt.get("value") == value:
                return opt.get("label", str(value))
        return str(value)
    if qtype == "multi_choice" and isinstance(value, list):
        labels = []
        by_value = {o.get("value"): o.get("label") for o in question.get("options", [])}
        for v in value:
            labels.append(by_value.get(v, str(v)))
        return ", ".join(labels) if labels else "—"
    return str(value)
